README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# unhcrdataportal

<!-- badges: start -->

[![R-CMD-check](https://github.com/Edouard-Legoupil/unhcrdataportal/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/Edouard-Legoupil/unhcrdataportal/actions/workflows/R-CMD-check.yaml)
<!-- badges: end -->

The goal of unhcrdataportal is to retrieve data from UNHCR Data Portal
within an R session.

Retrieving this data allows to perform quick quality assurance control.

## Installation

You can install the development version of unhcrdataportal from
[GitHub](https://github.com/) with:

``` r
# install.packages("pak")
pak::pkg_install("Edouard-Legoupil/unhcrdataportal")
```

Note that if you are on windows, you will need to have Rtools installed
on your machine

## Example

This is a basic example

``` r
library(unhcrdataportal)
## basic example code
page = "Sudan situation"
datasitpop <- odp(page = page,
               dataset = "population") |> 
  ## extract the data
  purrr::map( "data") |>
  ## Bind together in a data frame
  purrr::list_rbind()

library(ggplot2)
datasitpop |>
  ggplot() +
  geom_col(aes(x = as.integer(individuals),
               y = reorder( glue::glue("{population_group_name} \n as of {date}"), 
                                        as.integer(individuals) )),
           fill = unhcrthemes::unhcr_pal(n = 1, "pal_blue"),
           width = 0.8) +
  labs(title = paste0(page),
       x = "Number of people",
       y = "",
       caption = "Source: https://data.unhcr.org/en/situations \n
                   © UNHCR, The UN Refugee Agency") +
  scale_x_continuous(expand = expansion(c(0, 0.1)),
                     breaks = scales::pretty_breaks(n = 7),
                     labels = scales::label_number(scale_cut = scales::cut_short_scale())) +
  unhcrthemes::theme_unhcr(
              font_size = 12,
              grid = "X",
              axis = "y")
```

<img src="man/figures/README-example-1.png" width="100%" />


NEWS.md
# unhcrdataportal 0.0.0.9000

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
library(testthat)
library(unhcrdataportal)

test_check("unhcrdataportal")


tests/testthat/test-odp.R
# WARNING - Generated by {fusen} from /dev/flat_full.Rmd: do not edit by hand


test_that("odp works", {
  expect_true(inherits(odp, "function")) 
})

# Define unit tests for odp function
test_that("odp returns sitemap when page is NULL", {
  sitemap <- odp()
  expect_true(is.data.frame(sitemap))
})

test_that("odp returns specific page from sitemap", {
  page <- "Sudan situation"
  result <- odp(page)
  expect_true(is.data.frame(result))
  expect_equal(result[["page_name"]][[1]], page)
})


test_that("odp returns population data correctly", {
  page <- "Sudan situation"
  dataset <- "population"
  result <- odp(page = page, dataset =  dataset  )
  expect_true(is.list(result))
})




tests/testthat/test-run_app.R
# WARNING - Generated by {fusen} from /dev/flat_full.Rmd: do not edit by hand

test_that("run_app works", {
  expect_true(inherits(run_app, "function")) 
})


R/run_app.R
# WARNING - Generated by {fusen} from /dev/flat_full.Rmd: do not edit by hand

#' run_app
#' 
#' 
#' A quick shiny App to show the data the function is extracting
#' 
#' @import shiny
#' @import ggplot2
#' @importFrom unhcrthemes unhcr_pal theme_unhcr
#' @importFrom DT DTOutput
#' @importFrom stats reorder
#' @importFrom glue glue
#' @importFrom dplyr pull filter
#' @importFrom purrr map list_rbind
#' @importFrom scales pretty_breaks label_number cut_short_scale
#' 
#' @return nada
#' 
#' @export
#' @examples
#' 
#' ## Run the shiny App 
#' # run_app()
run_app <- function(){
ui <- fluidPage(
  h1("Retrieve Data From UNHCR Data Portal"),
  selectInput(inputId= "liste", 
              label = " then select a situation",
              choices = 
                unhcrdataportal::odp() |> 
                dplyr::filter(page_type =="situation" ) |>
                dplyr::pull(page_name) ),
                
              #  NULL ),
  
  actionButton("go", "Pull Data") ,
  tabsetPanel(type = "tabs",
              tabPanel("Plot",
                       plotOutput(outputId ="chart"),
                       " " ),
              tabPanel("Table",
                       DT::DTOutput(outputId = "table"),
                       " ")
              ),
  p("source code in package :",
    tags$a(href="https://github.com/Edouard-Legoupil/unhcrdataportal", "unhcrdataportal") )
    
)

server <- function(input, output, session) {
  ## create a reactive value to store what we have 
  reactiveObject <- reactiveValues(
    data = NULL,
    plot = NULL  )
  observeEvent(eventExpr = input$go, {
    data <- odp(page = as.character(input$liste),
                dataset = "population") |>
      ## extract the data
      purrr::map( "data") |>
      ## Bind together in a data frame
      purrr::list_rbind() 
    reactiveObject$data  <- data
    
    plot <- data |>
      ggplot() +
      geom_col(aes(x = as.integer(individuals),
                   y = stats::reorder( glue::glue("{population_group_name} \n as of {date}"),
                                       as.integer(individuals) )),
               fill = unhcrthemes::unhcr_pal(n = 1, "pal_blue"),
               width = 0.8) +
      labs(title = paste0(as.character(input$liste)),
           x = "Number of people",
           y = "",
           caption = "Source: UNHCR Data Portal: https://data.unhcr.org/en/situations \n\u00a9 UNHCR, The UN Refugee Agency") +
      scale_x_continuous(expand = expansion(c(0, 0.1)),
                         breaks = scales::pretty_breaks(n = 7),
                         labels = scales::label_number(scale_cut = scales::cut_short_scale())) +
      unhcrthemes::theme_unhcr( font_size = 22, grid = "X",  axis = "y")
    
    reactiveObject$plot  <- plot
    
    ## console output in case..
    # cat(file=stderr(), 
    #     " List sit", 
    #     allsit ,"\n")
    
  })
  ## generate the plot
  output$chart <- renderPlot({
       reactiveObject$plot
    })
  ## generate the table
  output$table <- DT::renderDT({
           reactiveObject$data
    })
}

shinyApp(ui, server)
}


R/odp.R
# WARNING - Generated by {fusen} from /dev/flat_full.Rmd: do not edit by hand

#' Retrieve data from UNHCR Operational Data Portal 
#' 
#' 
#' This function pulls key information from the portal
#' 
#' @param page aka the name of the situation
#' @param dataset one of the following: "population", "population_x_time",  
#'                 "population_x_location",  "population_x_origin", 
#'                  "population_x_location2",  "demographics",  "demographics" 
#' 
#' @return a frame with data
#' 
#' @references \href{https://data.unhcr.org/}{UNHCR Operational Data Portal}
#' 
#' @importFrom glue glue
#' @importFrom rvest html_elements html_attr session
#' @importFrom purrr map list_rbind modify_tree  map_chr
#' @importFrom tibble tibble as_tibble
#' @importFrom httr modify_url
#' @importFrom dplyr mutate case_when filter
#' @importFrom stringr str_detect str_match
#' @importFrom tidyr unnest 
#' @importFrom jsonlite fromJSON
#' 
#' @export
#' @examples
#' ## Use as
#' 
#' ## odp(): to retrieve an overview of all available situation pages
#' data <- odp() |> 
#'          dplyr::filter(page_type =="situation" )
#' 
#' DT::datatable(data |> dplyr::select( page_name,page_url),
#'               options = list(scrollX = TRUE))
#' 
#' ## to retrieve the datasets for a specific country /situation
#'  
#' datasit <- odp(page = "Sudan situation")
#' # display results.. 
#' DT::datatable(datasit,
#'               options = list(scrollX = TRUE))
#' 
#' ## to retrieve a specific dataset on the given page as a parsed JSON
#' # odp(page, dataset)
#' page <- "Sudan situation"
#' dataset <- "population"
#' 
#' datasitpop <- odp(page = page,
#'                dataset = dataset) |> 
#'   ## extract the data
#'   purrr::map( "data") |>
#'   ## Bind together in a data frame
#'   purrr::list_rbind()
#' 
#' ## plotting quickly the results... 
#' library(ggplot2)
#' datasitpop |>
#'   ggplot() +
#'   geom_col(aes(x = as.integer(individuals),
#'                y = reorder( glue::glue("{population_group_name} \n as of {date}"), 
#'                                         as.integer(individuals) )),
#'            fill = unhcrthemes::unhcr_pal(n = 1, "pal_blue"),
#'            width = 0.8) +
#'   labs(title = paste0(page),
#'        x = "Number of people",
#'        y = "",
#'        caption = "Source: https://data.unhcr.org/en/situations \n
#'                    © UNHCR, The UN Refugee Agency") +
#'   scale_x_continuous(expand = expansion(c(0, 0.1)),
#'                      breaks = scales::pretty_breaks(n = 7),
#'                      labels = scales::label_number(scale_cut = scales::cut_short_scale())) +
#'   unhcrthemes::theme_unhcr(
#'               font_size = 12,
#'               grid = "X",
#'               axis = "y")
odp <- function(page = NULL, dataset = NULL) {
  
  # Step 1: Check if the .odp environment variable exists in the global environment,
  # and create it if it doesn't exist.
  if (is.null(.GlobalEnv$.odp))
    .GlobalEnv$.odp <- new.env(parent = emptyenv())

  
  # Step 2: Check if the sitemap exists in the .odp environment variable,
  # and create it if it doesn't exist.
  if (is.null(.GlobalEnv$.odp$sitemap))
    .GlobalEnv$.odp$sitemap <- 
      rvest::session(glue::glue("https://data.unhcr.org/en/site-map/")) |> 
      rvest::html_elements("a[href*='/situations/'], 
                     a[href*='/country/']") |> 
      purrr::map(\(x) tibble::tibble(page_name = rvest::html_attr(x, "title"),
                      page_url = httr::modify_url("https://data.unhcr.org/",
                                                  path = rvest::html_attr(x, "href")))) |> 
      purrr::list_rbind() |> 
      dplyr::mutate(page_type = dplyr::if_else(stringr::str_detect(page_url, "/country/"), 
                                               "country",
                                               "situation"),
             .before = page_name)
  
  # Step 3: Return the sitemap if the page argument is NULL.
  if (is.null(page))
    return(.GlobalEnv$.odp$sitemap)
  
  # Step 4: Check if the requested page exists in the .odp environment variable,
  # and create it if it doesn't exist.
  if (is.null(.GlobalEnv$.odp[[page]]))
    .GlobalEnv$.odp[[page]] <- 
      .GlobalEnv$.odp$sitemap |> 
      dplyr::filter(page_name == page) |>
      dplyr::mutate(datasets = purrr::map(page_url, 
                            \(url) rvest::session(url) |> 
                              rvest::html_elements(xpath = "//div[a[contains(text(), \'JSON\')]]") |> 
                              (\(x) tibble::tibble(dataset_name = x |> 
                                             rvest::html_text(trim = TRUE) |> 
                                             (\(x) stringr::str_match(x, "(.+)\\n")[,2])(),
                                           dataset_url = x |> 
                                             rvest::html_element("a:contains(\'JSON\')") |> 
                                             rvest::html_attr("href")))() |> 
                              dplyr::filter(dataset_url != "#"))) |> 
      tidyr::unnest(datasets) |> 
      dplyr::mutate(dataset_type = purrr::map_chr(dataset_url, \(x) httr::parse_url(x)$path),
             dataset_type = dplyr::case_when(
               dataset_type == "population/" ~ "population",
               dataset_type == "population/get/timeseries" ~ "population_x_time",
               dataset_type == "population/get/sublocation" ~ "population_x_location",
               dataset_type == "population/get/origin" ~ "population_x_origin",
               dataset_type == "population/get/sublocation/root" ~ "population_x_location2",
               dataset_type == "population/get/demography" ~ "demographics",
               dataset_type == "population/get/demography/simplified" ~ "demographics"),
             .before = dataset_name)
  
  # Step 5: Return something if dataset variable was not documented..
  if (is.null(dataset))  
    
    return(.GlobalEnv$.odp[[page]])

  if (is.null(.GlobalEnv$.odp[[glue::glue("{page}/{dataset}")]]))  
  # Check if the requested page/dataset combination exists in the .odp environment variable,
  # and create it if it doesn't exist.
    
    .GlobalEnv$.odp[[glue::glue("{page}/{dataset}")]] <- 
      ## fixed below...jsonlite::fromJSON takes on single argument ...
     # jsonlite::fromJSON(filter(.GlobalEnv$.odp[[page]],
      ## Fixed below dataset_type instead of dataset_name
      # dataset_name == dataset)$dataset_url) |> 
      dplyr::filter(.GlobalEnv$.odp[[page]], 
                    dataset_type == dataset)$dataset_url |>
      purrr::map (jsonlite::fromJSON)  |> 
      purrr::modify_tree(leaf = \(x) if (is.data.frame(x)) tibble::as_tibble(x) else x)
  
  #  Return the dataset for the requested page/dataset combination.
    return(.GlobalEnv$.odp[[glue::glue("{page}/{dataset}")]] ) 
   
}
